org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/SwitchExpression.java

```diff
@@ -341,31 +344,26 @@ public TypeBinding resolveType(BlockScope upperScope) {
 					return this.resolvedType = null;
 				}
 
-				if (this.originalValueResultExpressionTypes == null) {
-					this.originalValueResultExpressionTypes = new TypeBinding[resultExpressionsCount];
-					this.finalValueResultExpressionTypes = new TypeBinding[resultExpressionsCount];
-					for (int i = 0; i < resultExpressionsCount; ++i) {
-						this.finalValueResultExpressionTypes[i] = this.originalValueResultExpressionTypes[i] =
-								this.resultExpressions.get(i).resolvedType;
-					}
+				this.originalValueResultExpressionTypes = new TypeBinding[resultExpressionsCount];
+				this.finalValueResultExpressionTypes = new TypeBinding[resultExpressionsCount];
+				for (int i = 0; i < resultExpressionsCount; ++i) {
+					this.finalValueResultExpressionTypes[i] = this.originalValueResultExpressionTypes[i] = this.resultExpressions.get(i).resolvedType;
 				}
+
 				if (isPolyExpression()) { //The type of a poly switch expression is the same as its target type.
 					if (this.expectedType == null || !this.expectedType.isProperType(true)) {
 						return new PolyTypeBinding(this);
 					}
 					return this.resolvedType = computeConversions(this.scope, this.expectedType) ? this.expectedType : null;
 				}
-			} else {
-				// re-resolving of poly expression:
+			} else { // re-resolving of poly expression:
 				resultExpressionsCount = this.resultExpressions.size();
 				if (resultExpressionsCount == 0)
 					return this.resolvedType = null; // error flagging would have been done during the earlier phase.
 				boolean yieldErrors = false;
 				for (int i = 0; i < resultExpressionsCount; i++) {
 					Expression resultExpr = this.resultExpressions.get(i);
-					TypeBinding origType = this.originalTypeMap.get(resultExpr);
-					// NB: if origType == null we assume that initial resolving failed hard, rendering re-resolving impossible
-					if (origType != null &&  origType.kind() == Binding.POLY_TYPE) {
```

Comment:
I'd need a good explanation why the previous fix was not good. It was quite expensive to achieve at.

Reply:
> I'd need a good explanation why the previous fix was not good. It was quite expensive to achieve at.

Hi Stephan, Just so we are clear, I did not say the fix for https://github.com/eclipse-jdt/eclipse.jdt.core/issues/520 was "not good."

But it is not meaningful, if the code whose problem it is fixing is going away.

Let us summarize the picture:

- A switch expression is a poly expression when it features in an assignment or invocation context.
- During the initial resolve of a poly switch expression, target type may or may not be available (invocation - not available, assignment - available)
- When target type is not available in the initial resolve in invocation context, we return a PolyTypeBinding
- When target type becomes available after overload resolution, the switch needs be re-resolved against that target type.
- In this last step, it is only needed to re-resolve poly result expressions. Any standalone result expressions of the switch will not be influenced by the target type imposition and so there is no point in re-resolving them.
- So we need to ask which result expressions need to be re-resolved: For this the fix for `https://bugs.eclipse.org/bugs/show_bug.cgi?id=576861` invented a new mechanism using the newly introduced state `originalTypeMap` which needs to be maintained involving some new code.

I am saying no such new mechanism needs to be invented to map a result expression to be a binding and then querying the kind thusly: `kind() == Binding.POLY_TYPE`

You can simply ask the result expression (or any expression for that matter that has undergone an initial resolve) using the existing API: `org.eclipse.jdt.internal.compiler.ast.Expression.isPolyExpression()`

So this code on master:

```
                                        Expression resultExpr = this.resultExpressions.get(i);
					TypeBinding origType = this.originalTypeMap.get(resultExpr);
					// NB: if origType == null we assume that initial resolving failed hard, rendering re-resolving impossible
					if (origType != null &&  origType.kind() == Binding.POLY_TYPE) {
```
 
becomes 

```
                                        Expression resultExpr = this.resultExpressions.get(i);
                                        if (resultExpr.isPolyExpression()) {
```

How does that sound ?